### ISSUE TITLE
Save game corruption

### ISSUE THREAD

**Issue by pugwash99** [ryobg/JContainers#5]
A debug build doesn't run correctly but the mods I'm testing appear to work with a release build. The mod I'm testing with appears fully functional.

The problem appears when attempting to load a save game, either the load function or the save is corrupting the data and causing another ctd.

**Comment by ryobg** [comment]
Can you elaborate a bit more on "Debug build does not run correctly"? I havent look soon in the Debug configuration, but we need it running so that we can debug cases like this one :)

**Comment by pugwash99** [comment]
I think the debug issue was a complete mistake on my part, I was running with a debug session and for some reason it wasn't registering some of the data coming from a json table being read from file. Since I recompiled and ran without a debugger attached it's working correctly.

However, with a jcontainers debug build installed an old save without JC data in the skse save file will load correctly. Save the game and then reload I get an exception right at the moment where it should load data. The message in a pop-up window is

file: minkernal/crts\ucrt\src\appcrt\heap\debug_heap.cpp Line 888

Expression: _CrtlsValidHeapPointer(block)

Checking the skse counterpart file the json data is present.

With a VS debugging session attached to skyrim it results in a ctd without the pop-up. No break points work on the save or load functions in the skse_callbacks.cpp, so either the save is corrupting the data or the load is not working or reading past where it should.

**Comment by ryobg** [comment]
Did you narrowed it down? Tracing is enough to pinpoint such an issue.

**Comment by pugwash99** [comment]
A little bit, in routine  'auto read_from_stream(master& self, std::istream& stream) -> void {' in domain_master.cpp

I've a JC log before and after the line 'archive >> self;' the second doesn't appear in the log file,  It's around line 200

**Comment by ryobg** [comment]
Hmm not sure... Is there any way for me to reproduce that easily? I can only speculate otherwise. Maybe one of the numerous `load(Archive&, unsigned)` functions is failing down there.

**Comment by pugwash99** [comment]
As the original code was working I can't imagine it's something in the boost code. I did find one possible reason in another post.

If you serialize an object with a header that uses a 32bit variable but deserialize to a 64bit or vice versa it will also cause an exception. 

It could equally be that I've got old data hanging around in a skse and it's causing the problem as . I'll try deleting the skse file and see what happens.

**Comment by pugwash99** [comment]
well that's a bit annoying, I replicated the full folder tree from the original jcontainers which includes a dummy file in the domains folder. Delete the file and the crashes stop happening. It must be related to the newer domains code that was included in v3.3 because without that file I can save and load.

It might also explain why you may not be having the problem, if you only added the domains folder and not the dummy file as well.

**Comment by ryobg** [comment]
Can you check out whether this happen with the new distro? In the new one I have added one `.force-install` dummy file. It should be ignored during load of domains.

I was chasing some issues with the backward compatibility of loading archives. Certainly not. Boost does uses native sizes and the offsets get messed up everywhere.

**Comment by pugwash99** [comment]
Boost 32bit and boost 64bit have a known incompatibility with binary archives, text files are find but binary definitely appear have a 4 byte increase in size where a 32bit varible is being replaced by a 64bit one.  It was one of the things I was looking into.

I'll download it in a while and test it but without the dummy file I can definitely save and load without a problem. Add the file back and the skse save file must reference it in some way causing a ctd in the deserializing processes.

**Comment by ryobg** [comment]
Let me know the importance of supporting reads of old archives. As I said, the issue is built inside Boost. Small example: when you need to read a string back, Boost reads now 8 bytes for its size instead of 4. As you can imagine, it tries then to allocate some huge storage and you get an exception if you are lucky. I can imagine possible solutions but these will be PITA. 

I want to know whether the issue with that dummy file persist. I will check again too of course.

**Comment by pugwash99** [comment]
Personally it's not important, True it looses some data but it might be better to start clean as when other mods start to load data it could cause more problems.

I've rebuilt with the fresh release, I loaded and saved a few times, moved location, saved again, closed skyrim, reopened and then reloaded no ctd. There's no mention of the .forced install file that's in the domains folder and loading a save made with the version I was working with didn't crash but it did freeze for a few seconds.
